core/src/test/java/org/apache/calcite/test/catalog/Fixture.java

```diff
@@ -109,16 +109,23 @@
             .build();
     empListType =
         typeFactory.createArrayType(empRecordType, -1);
+
+    // Subclass ObjectSqlType to make nested field queryable.
     addressType =
-        new ObjectSqlType(SqlTypeName.STRUCTURED,
+        new ObjectSqlType(
+            SqlTypeName.STRUCTURED,
```

Comment:
Changing only the BookStore schema should be sufficient to demonstrate what you want. In RelToSqlConverterTest you should only pass the appropriate SchemaSpec before writting your sql.

Reply:
Hey @zabetak, I tried this Bookstore schema in SqlToRelConverterTest. The change is much bigger than the original one because to add Bookstore schema in SqlToRelTestBase takes:
1. Change TesterImpl type factory to JavaTypeFactoryImpl -> add withTypeFactory method in both Tester interface and impl.
2. Do the same thing for RelOptPlanner in TesterImpl and replace it with a subclass of MockRelOptPlanner with traitset related method overrided. 
3. To support the Bookstore schema, we need make the CalciteSchema to be cached instead of existing simple, and make it case insensitive -> change MockCatalogReader constructor signature and subclass it to register bookstore schema.

I think there is no need that we stick to Bookstore because there is already existing nested fields in MockCatalogReaderSimple. Another approach is to adapt the existing changes on Bookstore schema to MockCatalogReaderSimple (basically add one more layer deeper in a nested schema). And migrate PlannerTest cases to SqlToRelConverterTest. IMO this approach is cleanest and also aligned with the existing tests.

What do you think?

Reply:
@zabetak And if you think OK, we could just keep the existing second-level-access test I implemented in SqlToRelConverterTest. I don't think there is much difference between the second level and the third level.

If so, I will delete all changes made to Bookstore and PlannerTest.

let me know.

Reply:
Thanks for your detailed explanation @my7ym! Indeed, I was not expecting it to be so complicated. I agree with you, probably the best place is SqlToRelConverterTest and there is no need to touch the Bookstore schema.

Reply:
No problem. Learnt a lot actually and more familiar with the code base :). Pls take another look man. Thanks.